Add support for macOS brew

tuedachu opened the pull request:
First version of 'helm-system-package' for home brew 

Working functionalities:
- Show information
- Install/Uninstall
- Browse URL
- Link/Unlink app

TODO:
(1) Implement `show-dependencies' function
(2) Show packages explicitly installed (function `helm-system-packages-brew-list-explicit' already exists)
(3) Implement `find-files'
(4) Keymap

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+;;; helm-system-packages-brew.el --- Helm UI for Mac OS ' homebrew. -*- lexical-binding: t -*-
```
The OS name is officially "macOS".  Also no space before the apostrophe.

piotr-yuxuan replied:
Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;; Helm UI for Mac OS' homebrew.
```
"macOS".

piotr-yuxuan replied:
Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+					   (format-string (format "%%-%ds  %%s" helm-system-packages-column-width)))
+				    (format format-string name desc)))
+				  (split-string descriptions"\n") "\n" ))
+  (cons names descriptions)))
```
The cache construction is a little different now, you should use `helm-system-packages--cache-set`.  See -pacman.el or -xbps.el.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+      'global
+    (if helm-system-packages-show-descriptions-p
+        helm-system-packages--descriptions
+      helm-system-packages--names)))
```
This function can now be removed altogether: use `helm-system-packages-init` instead.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+(defun helm-system-packages-brew-refresh ()
+  "Refresh the package list."
+  (interactive)
+  (setq helm-system-packages--source-name "brew source")
```
The title is now stored in the cache.  See `helm-system-packages--cache`.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+    (let ((res (helm-system-packages-brew-cache)))
+      (setq helm-system-packages--names (car res)
+            helm-system-packages--descriptions (cdr res)))
+    (setq helm-system-packages--display-lists nil))
```
Display lists are important: they allow for fontifying the different categories (installed packages, dependencies, orphans, etc.).  Again, see `helm-system-packages-(xbps|pacman)-refresh`.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+  (setq helm-system-packages-column-width
+        (or helm-system-packages-brew-column-width
+            helm-system-packages-column-width))
+    (let ((res (helm-system-packages-brew-cache)))
```
The caching mechanism has changed, as stated previously.  Refer to the refresh functions of XBPS or pacman for an example.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+	 str
+	 pkg
+	 (i 0))
+    (dolist (pkg (helm-marked-candidates) nil)
```
`nil` is unnecessary here.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+	      "\n\n"
+	      "* Caveats: " (alist-get 'caveats pkg-desc-alist) "\n"))
+     (add-to-list 'desc-list `(uninstalled (,pkg . ,str)))
+     (setq i (+ i 1)))
```
Use the `1+` function.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+
+(defun helm-system-package-brew-unlink-app (_candidate)
+  (helm-system-packages-brew-run  "brew" "unlink"))
+
```
No need for multiple line breaks, I think.

piotr-yuxuan replied:
Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+   (let* ((descriptions (json-read-from-string (with-temp-buffer
+						(apply 'call-process "brew" nil t nil "info" "--json=v1" (helm-marked-candidates)) 
+						(buffer-string)))))
+    (helm-system-packages-browse-url (mapcar (lambda(pkg)
```
Missing space before `(pkg)`.

piotr-yuxuan replied:
Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+
+
+
+(defun helm-system-packages-brew-run (command &rest args)
```
Just to be sure: This is the same as `helm-system-packages-call-as-root` except that it does not use `sudo`, right?
If so, mention it in the documentation.  Also `helm-system-packages-call-as-root` has undergone some changes, you might want to reflect them here.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+     (lambda (_)
+       (helm-system-packages-brew-run "brew" "uninstall"
+                                         (when helm-current-prefix-arg "--force"))))
+    
```
No linebreak here.

piotr-yuxuan replied:
Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

Ambrevar commented on helm-system-packages-brew.el:
```diff
@@ -0,0 +1,214 @@
+
+(defun helm-system-packages-brew-build-source ()
+  "Build Helm source for brew."
+  (helm-build-in-buffer-source helm-system-packages--source-name
```
The way the source name is set has changed: see -xbps.el or -pacman.el.

Ambrevar reviewed (commented):
Looks good overall but your contribution went out of sync with master since I contributed the support for remote systems via TRAMP.  Bad timing, sorry!
Try to address my comments, test and repush, I'll merge then.

Let me know if you run into any issue.  Feel free to ask for further explanations regarding the new cache system.

Commit:
brew: Address minor PR remarks

Minor remarks from PR
https://github.com/emacs-helm/helm-system-packages/pull/16#pullrequestreview-104919276
are addressed:

https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175399186
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175399290
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175402037
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175402208
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175403026
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175403100
https://github.com/emacs-helm/helm-system-packages/pull/16#discussion_r175405036

tuedachu committed:
brew: Init

tuedachu committed:
readme: Remove 'brew' from the roadmap

Commit:
brew: Upgrade architecture to meet 'helm-system-packages' requirements

Commit:
helm-system-packages: Add 'brew' in the list of package managers
